packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenModule.m

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018-present 650 Industries. All rights reserved.
+
+#import <EXSplashScreen/EXSplashScreenModule.h>
+#import <EXSplashScreen/EXSplashScreenService.h>
+#import <UMCore/UMAppLifecycleService.h>
+#import <UMCore/UMUtilities.h>
+
+@interface EXSplashScreenModule ()
+
+@property (nonatomic, weak) UMModuleRegistry *moduleRegistry;
+@property (nonatomic, weak) id<UMUtilitiesInterface> utilities;
+
+@end
+
+@implementation EXSplashScreenModule
+
+UM_EXPORT_MODULE(ExpoSplashScreen);
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+{
+  _moduleRegistry = moduleRegistry;
+  _utilities = [moduleRegistry getModuleImplementingProtocol:@protocol(UMUtilitiesInterface)];
+  [[moduleRegistry getModuleImplementingProtocol:@protocol(UMAppLifecycleService)] registerAppLifecycleListener:self];
+}
+
+UM_EXPORT_METHOD_AS(hideAsync,
+                    hideWithResolve:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  UM_WEAKIFY(self);
+  dispatch_async(dispatch_get_main_queue(), ^{
+    UM_ENSURE_STRONGIFY(self);
+    UIViewController* currentViewController = self.utilities.currentViewController;
+    
+    if (!currentViewController) {
+      reject(@"ERR_SPLASH_SCREEN", @"No valid ViewController.", nil);
+    }
+    
+    [[self splashScreenService] hide:currentViewController
+                     successCallback:^{ resolve(nil); }
+                     failureCallback:^(NSString *message){ reject(@"ERR_SPLASH_SCREEN", message, nil); }];
+  });
+}
+
+UM_EXPORT_METHOD_AS(preventAutoHideAsync,
+                    preventAutoHideWithResolve:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  UM_WEAKIFY(self);
+  dispatch_async(dispatch_get_main_queue(), ^{
+    UM_ENSURE_STRONGIFY(self);
+    UIViewController* currentViewController = self.utilities.currentViewController;
+
+    if (!currentViewController) {
+      reject(@"ERR_SPLASH_SCREEN", @"No valid ViewController.", nil);
+    }
+
+    [[self splashScreenService] preventAutoHide:currentViewController
+                                successCallback:^{ resolve(nil); }
+                                failureCallback:^(NSString *message){ reject(@"ERR_SPLASH_SCREEN", message, nil); }];
+  });
+}
+
+# pragma mark - UMAppLifecycleListener
+
+- (void)onAppBackgrounded {}
+
+- (void)onAppForegrounded {}
+
+- (void)onAppContentDidAppear
+{
+  UM_WEAKIFY(self);
+  dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Can it be `async`?

Reply:
It actually is `async` 🤔 ❓

Reply:
Let me rephrase:

Isn't this a problem that we call `onAppContentDidAppear` asynchronously? Shouldn't we call it synchronously?

Reply:
🤔 Nope, that's not a problem - we'll just have that action triggered a few moments later (and this action is just reshowing SplashScreen).

Reply:
Hm, what I was afraid of was that eg. there is a visible flash between when a white reload screen is shown and the time when block scheduled in `onAppContentWillReload` shows the splash screen.